components/support/sql/src/open_database.rs

```diff
@@ -0,0 +1,432 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+/// Use this module to open a new SQLite database connection.
+///
+/// Usage:
+///    - Define a struct that implements MigrationLogic.  This handles:
+///      - Initializing the schema for a new database
+///      - Upgrading the schema for an existing database
+///      - Extra preparation/finishing steps, for example setting up SQLite functions
+///
+///    - Call open_database() in your database constructor:
+///      - The first method called is setup_pragmas().  This is executed outside of a transaction
+///        to support `PRAGMA journal_mode=wal`
+///      - If the database file is not present, open_database() will create a new DB and call prepare(),
+///        init(), then finish()
+///      - If the database file exists, open_database() will open it and call prepare(),
+///        upgrade_from() for each upgrade that needs to be applied, then finish().
+///
+///  See the autofill DB code for an example.
+///
+use crate::ConnExt;
+use rusqlite::{Connection, OpenFlags, NO_PARAMS};
+use std::path::Path;
+use thiserror::Error;
+
+#[derive(Error, Debug)]
+pub enum Error {
+    // Generic error meaning that something went wrong during the migration.  You can return this
+    // from your upgrade functions to signal that the database is beyond repair and can't be
+    // migrated.
+    #[error("MigrationError: {0}")]
```

Comment:
I though it could be used by the upgrade_from() function if it detected some issue with the data it couldn't recover from.  But I'm thinking that's pretty unlikely.  I deleted that variant and a few others.